src/dali/abstract_ccxt_input_plugin.py

```diff
@@ -0,0 +1,651 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# CCXT documentation:
+# https://docs.ccxt.com/en/latest/index.html
+
+# pylint: disable=fixme
+
+import json
+import logging
+from datetime import datetime, timezone
+from multiprocessing.pool import ThreadPool
+from typing import Any, Dict, List, NamedTuple, Optional, Union
+
+from ccxt import Exchange
+from rp2.logger import create_logger
+from rp2.rp2_decimal import ZERO, RP2Decimal
+
+from dali.abstract_input_plugin import AbstractInputPlugin
+from dali.abstract_transaction import AbstractTransaction
+from dali.ccxt_pagination import (
+    AbstractPaginationDetailSet,
+    AbstractPaginationDetailsIterator,
+    PaginationDetails,
+)
+from dali.configuration import Keyword
+from dali.in_transaction import InTransaction
+from dali.intra_transaction import IntraTransaction
+from dali.out_transaction import OutTransaction
+
+# Native format keywords
+_AMOUNT: str = "amount"
+_BUY: str = "buy"
+_COST: str = "cost"
+_CURRENCY: str = "currency"
+_DATE_TIME: str = "datetime"
+_DEPOSIT: str = "deposit"
+_FEE: str = "fee"
+_FETCH_DEPOSITS: str = "fetchDeposits"
+_FETCH_MY_TRADES: str = "fetchMyTrades"
+_FETCH_WITHDRAWALS: str = "fetchWithdrawals"
+_ID: str = "id"
+_PRICE: str = "price"
+_SELL: str = "sell"
+_SIDE: str = "side"
+_STATUS: str = "status"
+_SYMBOL: str = "symbol"
+_TIMESTAMP: str = "timestamp"
+_TYPE: str = "type"
+_TX_ID: str = "txid"  # CCXT doesn't capitalize I
+_WITHDRAWAL: str = "withdrawal"
+
+_MS_IN_SECOND: int = 1000
+
+
+class Trade(NamedTuple):
+    base_asset: str
+    quote_asset: str
+    base_info: str
+    quote_info: str
+
+
+class ProcessOperationResult(NamedTuple):
+    in_transactions: List[InTransaction]
+    out_transactions: List[OutTransaction]
+    intra_transactions: List[IntraTransaction]
+
+
+class AbstractCcxtInputPlugin(AbstractInputPlugin):
+
+    __DEFAULT_THREAD_COUNT: int = 1
+
+    def __init__(
+        self,
+        account_holder: str,
+        exchange_start_time: datetime,
+        native_fiat: str,
+        thread_count: Optional[int],
+    ) -> None:
+
+        super().__init__(account_holder, native_fiat)
+        self._logger: logging.Logger = create_logger(f"{self.exchange_name()}/{self.account_holder}")
+        self.__cache_key: str = f"{str(self.exchange_name).lower()}-{account_holder}"
+        self._client: Exchange = self._initialize_client()
+        self._thread_count = thread_count if thread_count else self.__DEFAULT_THREAD_COUNT
```

Comment:
All fields still need to start with double underscore (our data models are encapsulated and we never access fields directly outside the class in which they are defined). To implement protected behavior they need a property that starts with single underscore (to access in the subclass). Sorry if my description wasn't clear when we discussed this.

Here's an updated text that's hopefully clearer:
* all class fields have two leading underscores because our data models are encapsulated and we never access fields directly outside the class in which they are defined
* private fields: visible only to the class, nothing else to do (access within the class via the double-underscore filed)
* protected fields: visible to the class and to its subclasses, needs to be encapsulated with a property or an accessor function that starts with one leading underscore 
* public fields: visible to everybody, needs to be encapsulated with a property or acccessor function that starts with no leading underscore.

Reply:
So, actual fields have leading double underscore:
```
self.__client
self.__thread_count
```

Then:
* if we want private access (this class only) we do nothing else
* if we want protected access (this class and its subclasses), we add a property with a single underscore (or an accessor method starting with `_get`)
* if we want public access (everybody), we add a property with no underscore (or an accessor method starting with `get`)

Whether to use a property or a get function is up to you: I tend to use properties if the access logic is fairly simple and get functions if access logic is more complicated. But again, either way works. Hope this is clearer, sorry if I didn't articulate it clearly enough earlier.

Reply:
I added a data encapsulation paragraph (with the above description) to the Design Guidelines section of README.dev.md both in RP2 and DaLI. Let me know if you think it needs more explanation: I'd like to fix the docs so that nobody trips on it in the future.